src/poetry/installation/wheel_installer.py

```diff
@@ -66,12 +87,52 @@ def for_source(self, source: WheelFile) -> WheelDestination:
         scheme_dict["headers"] = str(Path(scheme_dict["headers"]) / source.distribution)
 
         return self.__class__(
-            scheme_dict,
+            source=source,
+            scheme_dict=scheme_dict,
             interpreter=self.interpreter,
             script_kind=self.script_kind,
             bytecode_optimization_levels=self.bytecode_optimization_levels,
         )
 
+    def _validate_hash_and_size(
+        self, records: Iterable[tuple[Scheme, RecordEntry]]
+    ) -> None:
+        record_lines = self._source.read_dist_info("RECORD").splitlines()
+        record_mapping = {
+            record[0]: record for record in parse_record_file(record_lines)
+        }
+        for item in self._source._zipfile.infolist():
+            record_args = record_mapping.pop(item.filename, None)
+            if not record_args:
+                continue
+
+            file_record = RecordEntry.from_elements(*record_args)
+            computed_record = next(
+                record for _, record in records if record.path == item.filename
+            )
+            if (
+                file_record.hash_ is not None
+                and computed_record.hash_ is not None
+                and file_record.hash_ != computed_record.hash_
+            ) or (
+                file_record.size is not None
+                and computed_record.size is not None
+                and file_record.size != computed_record.size
+            ):
+                self.issues.append(
+                    f"In {self._source._zipfile.filename}, hash / size of"
+                    f" {item.filename} didn't match RECORD"
+                )
+
+    def finalize_installation(
+        self,
+        scheme: Scheme,
+        record_file_path: str,
+        records: Iterable[tuple[Scheme, RecordEntry]],
+    ) -> None:
+        self._validate_hash_and_size(records)
+        return super().finalize_installation(scheme, record_file_path, records)
```

Comment:
```suggestion
        try:
            self._validate_hash_and_size(records)
        finally:
            return super().finalize_installation(scheme, record_file_path, records)
```

I wonder if we should do something like that to make sure the `RECORD` file is written despite of bugs in `_validate_hash_and_size()`. However, when I tried this, the exception was swallowed, which not good either. Thus, we would have to put more effort here...